qiskit_ibm_runtime/execution_span.py

```diff
@@ -0,0 +1,141 @@
+from datetime import datetime
+from dataclasses import dataclass
+
+
+SliceType = Tuple[int, int]
+
+
+@dataclass(frozen=True)
+class ExecutionSpan:
+    """Stores an execution time span for a subset of job data."""
+
+    start: datetime
+    """The start time of the span, in UTC."""
+
+    stop: datetime
+    """The stop time of the span, in UTC."""
+
+    data_slices: Dict[int, SliceType]
+    r"""Which data have dependence on this execution span.
+
+    Data from the primitives are array-based, with every field in a 
+    :class:`~PubResult`\s :class:`~.DataBin` sharing the same base shape.
+    Therefore, the format of this field is a mapping from pub indexes to 
+    the same format accepted by 
+    NumPy slicing, where each value indicates which slice of each field in the
+    data bin depend on raw data collected during this execution span.
+    """
+
+    @property
+    def duration(self) -> float:
+        """Return the duration"""
+        return (self.stop - self.start).total_seconds()
+
+    def contains_pub(self, pub_idx: Union[int, Iterable[int]]) -> bool:
+        """Returns if a pub is contained"""
+        pub_idx = {pub_idx} if isinstance(pub_idx, int) else set(pub_idx)
+        return not pub_idx.isdisjoint(self.data_slices)
+
+    def filter_by_pub(self, pub_idx: Union[int, Iterable[int]]) -> "ExecutionSpan":
+        """Returns an ExecutionSpan filtered by pub"""
+        pub_idx = {pub_idx} if isinstance(pub_idx, int) else set(pub_idx)
+        slices = {idx: sl for idx, sl in self.data_slices.items() if idx in pub_idx}
+        return ExecutionSpan(self.start, self.stop, slices)
+
+    def to_tuple(self) -> Tuple:
+        """Return span in the form of a tuple"""
+        return (self.start, self.stop, self.data_slices)
+
+    @classmethod
+    def from_tuple(cls, span_tuple: Tuple) -> "ExecutionSpan":
+        """Construct a span from a tuple"""
+        return ExecutionSpan(*list(span_tuple))
+
+    def __str__(self) -> str:
+        return str(self.to_tuple())
+
+
+class ExecutionSpanSet:
+    """A collection of timings for the PUB result."""
+
+    def __init__(self, spans: Sequence[ExecutionSpan]):
+        self._spans = spans
+
+    def __len__(self) -> int:
+        return len(self._spans)
+
+    @overload
+    def __getitem__(self, idxs: int) -> ExecutionSpan: ...
+
+    @overload
+    def __getitem__(self, idxs: Union[slice, List[int]]) -> "ExecutionSpanSet": ...
+
+    def __getitem__(
+        self, idxs: Union[int, slice, List[int]]
+    ) -> Union[ExecutionSpan, "ExecutionSpanSet"]:
+        if isinstance(idxs, int):
+            return self._spans[idxs]
+        if isinstance(idxs, slice):
+            return ExecutionSpanSet(self._spans[idxs])
+        return ExecutionSpanSet([self._spans[idx] for idx in idxs])
+
+    def __iter__(self) -> Iterator[ExecutionSpan]:
+        return iter(self._spans)
+
+    def __str__(self) -> str:
+        return str(self.to_list_of_tuples())
```

Comment:
If only implementing one of repr or str, I usually prefer repr because str calls repr when undefined.

```suggestion
    def __repr__(self) -> str:
        return f"ExecutionSpanSet({self._spans})"
```

Reply:
This suggestion only shows the addresses of the execution spans, not their content.

Reply:
Since this PR implements `ExecutionSpan` as a dataclass, I think that the dataclass repr (and the list repr) will propagate, and not show the id.